contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -5107,3 +5109,127 @@ get_bbf_pivot_tuplestore(const char 	*sourcetext,
 
 	return tupstore;
 }
+
+/*
+ * tsql_charindex()
+ * Implementation of T-SQL CHARINDEX function to find a substring within a string.
+ */
+Datum
+tsql_charindex_char(PG_FUNCTION_ARGS)
+{
+    text *expressionToFind;
+    text *expressionToSearch;
+    int32 start_location;
+    int result = 0;
+    Oid collation;
+
+    /* Handle NULL inputs */
+    if (PG_ARGISNULL(0) || PG_ARGISNULL(1))
+        PG_RETURN_NULL();
+
+    /* Get function arguments */
+    expressionToFind = PG_GETARG_TEXT_PP(0);
+    expressionToSearch = PG_GETARG_TEXT_PP(1);
+    start_location = PG_GETARG_INT32(2);
+    
+    /* Handle empty search pattern case */
+    if (VARSIZE_ANY_EXHDR(expressionToFind) == 0)
+        PG_RETURN_INT32(0);
+    
+    /* Get collation from function call */
+    collation = PG_GET_COLLATION();
+
+    /* If no collation specified, use default */
+    if (!OidIsValid(collation))
+        collation = DEFAULT_COLLATION_OID;
+    
+    if (start_location <= 0)
+    {
+        /* Search from the beginning of the string using textpos with collation */
+        result = DatumGetInt32(DirectFunctionCall2Coll(textpos,
+                                                     collation,    
+                                                     PointerGetDatum(expressionToSearch),
+                                                     PointerGetDatum(expressionToFind)));
+    }
+    else
+    {
+        /* Get substring starting from start_location */
+        text *substr = DatumGetTextP(DirectFunctionCall2(text_substr_no_len,
+                                                      PointerGetDatum(expressionToSearch),
+                                                      Int32GetDatum(start_location)));
+        
+        /* Find position in substring with collation */
+        int pos = DatumGetInt32(DirectFunctionCall2Coll(textpos,
+                                                      collation,
+                                                      PointerGetDatum(substr),
+                                                      PointerGetDatum(expressionToFind)));
+        
+        /* Calculate final position if found */
+        if (pos > 0)
+            result = pos + start_location - 1;
+        
+        pfree(substr);
+    }
+    
+    PG_RETURN_INT32(result);
+}
+
+/*
+ * tsql_charindex_binary()
+ * Implementation of T-SQL CHARINDEX function for binary data types
+ */
+Datum
+tsql_charindex_binary(PG_FUNCTION_ARGS)
+{
+    bytea *expressionToFind;
+    bytea *expressionToSearch;
+    int32 start_location;
+    int result = 0;
+    
+    /* Check for NULL pattern and throw error */
+    if (PG_ARGISNULL(0))
+        ereport(ERROR,
+                (errcode(ERRCODE_NULL_VALUE_NOT_ALLOWED),
+                errmsg("Argument data type is invalid for argument 2 of charindex function.")));
```

Comment:
Why are we throwing invalid argument data type error when pattern is NULL, can you test following query with your changes and TSQL.
```
select charindex(cast(NULL as varbinary), 0x12345678)
go
```